File: stripe/src/test/java/com/stripe/android/view/CountryAdapterTest.java
package com.stripe.android.view;

import android.support.annotation.NonNull;

import androidx.test.core.app.ApplicationProvider;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;

import static org.junit.Assert.assertEquals;

/**
 * Test class for {@link CountryAdapter}
 */
@RunWith(RobolectricTestRunner.class)
public class CountryAdapterTest {

    private CountryAdapter mCountryAdapter;
    private List<String> mOrderedCountries;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        Locale.setDefault(Locale.US);

        mOrderedCountries = CountryUtils.getOrderedCountries(Locale.getDefault());
        mCountryAdapter = new CountryAdapter(
                ApplicationProvider.getApplicationContext(),
                mOrderedCountries
        );
    }

    @Test
    public void filter_whenEmptyConstraint_showsAllResults() {
        mCountryAdapter.getFilter().filter("");
        assertEquals(
                mOrderedCountries,
                getSuggestions()
        );
    }

    @Test
    public void filter_whenCountryInputNoMatch_showsAllResults() {
        mCountryAdapter.getFilter().filter("NONEXISTENT COUNTRY");
        assertEquals(
                mOrderedCountries,
                getSuggestions()
        );
    }

    @Test
    public void filter_whenCountryInputMatches_filters() {
        mCountryAdapter.getFilter().filter("United");
        assertEquals(
                Arrays.asList(
                        "United States",
                        "United Arab Emirates",
                        "United Kingdom",
                        "United States Minor Outlying Islands"
                ),
                getSuggestions()
        );
    }

    @Test
    public void filter_whenCountryInputMatchesExactly_showsAllResults() {
        mCountryAdapter.getFilter().filter("Uganda");
        assertEquals(
                mOrderedCountries,
                getSuggestions()
        );
    }

    @NonNull
    private List<String> getSuggestions() {
        final List<String> suggestions = new ArrayList<>(mCountryAdapter.getCount());
        for (int i = 0; i < mCountryAdapter.getCount(); i++) {
            suggestions.add(mCountryAdapter.getItem(i));
        }
        return suggestions;
    }
}

File: stripe/src/test/java/com/stripe/android/view/CountryAdapterTest.kt
package com.stripe.android.view

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import com.google.common.truth.Truth.assertThat
import java.util.Locale
import kotlin.test.BeforeTest
import kotlin.test.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

/**
 * Test class for [CountryAdapter]
 */
@RunWith(RobolectricTestRunner::class)
class CountryAdapterTest {

    private val countryAdapter: CountryAdapter by lazy {
        CountryAdapter(
            ApplicationProvider.getApplicationContext<Context>(),
            orderedCountries
        )
    }
    private val orderedCountries: List<Country> by lazy {
        CountryUtils.getOrderedCountries(Locale.getDefault())
    }

    private val suggestions: List<Country>
        get() {
            return (0 until countryAdapter.count).map {
                countryAdapter.getItem(it)
            }
        }

    @BeforeTest
    fun setup() {
        Locale.setDefault(Locale.US)
    }

    @Test
    fun filter_whenEmptyConstraint_showsAllResults() {
        countryAdapter.filter.filter("")
        assertThat(suggestions)
            .isEqualTo(orderedCountries)
    }

    @Test
    fun filter_whenCountryInputNoMatch_showsAllResults() {
        countryAdapter.filter.filter("NONEXISTENT COUNTRY")
        assertThat(suggestions)
            .isEqualTo(orderedCountries)
    }

    @Test
    fun filter_whenCountryInputMatches_filters() {
        countryAdapter.filter.filter("An")
        assertThat(suggestions.map { it.name })
            .isEqualTo(
                listOf(
                    "Andorra",
                    "Angola",
                    "Anguilla",
                    "Antarctica",
                    "Antigua and Barbuda"
                )
            )
    }

    @Test
    fun filter_whenCountryInputMatchesExactly_showsAllResults() {
        countryAdapter.filter.filter("Uganda")
        assertThat(suggestions)
            .isEqualTo(orderedCountries)
    }

    @Test
    fun updateUnfilteredCountries_withPopulatedSet_shouldUpdateSuggestions() {
        assertThat(countryAdapter.firstItem.code)
            .isEqualTo("US")
        assertThat(countryAdapter.updateUnfilteredCountries(setOf("fr", "de")))
            .isTrue()
        assertThat(countryAdapter.firstItem.code)
            .isEqualTo("FR")
    }

    @Test
    fun updateUnfilteredCountries_withEmptySet_shouldNotUpdateSuggestions() {
        assertThat(countryAdapter.firstItem.code)
            .isEqualTo("US")
        assertThat(countryAdapter.updateUnfilteredCountries(emptySet()))
            .isFalse()
        assertThat(countryAdapter.firstItem.code)
            .isEqualTo("US")
    }

    @Test
    fun updateUnfilteredCountries_shouldUpdateFilter() {
        countryAdapter.filter.filter("An")
        assertThat(suggestions.map { it.name })
            .isEqualTo(
                listOf(
                    "Andorra",
                    "Angola",
                    "Anguilla",
                    "Antarctica",
                    "Antigua and Barbuda"
                )
            )

        countryAdapter.updateUnfilteredCountries(setOf("ao", "ad"))
        countryAdapter.filter.filter("An")
        assertThat(suggestions.map { it.name })
            .isEqualTo(
                listOf(
                    "Andorra",
                    "Angola"
                )
            )
    }
}
